From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Maxlinear <nobody@maxlinear.com>
Date: Fri, 18 Feb 2022 10:00:33 +0100
Subject: Extend-vlan

act_vlan: allow only pcp or proto modify
The vlan modify can change the priority or the protocol without specifying
vlan vid value. This allows changing the VLAN pcp value for example.
---
 tc/m_vlan.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/tc/m_vlan.c b/tc/m_vlan.c
index 221083df..9419d45e 100644
--- a/tc/m_vlan.c
+++ b/tc/m_vlan.c
@@ -179,13 +179,21 @@ static int parse_vlan(struct action_util *a, int *argc_p, char ***argv_p,
 		}
 	}
 
-	if (has_push_attribs(action) && !id_set) {
+	if (action == TCA_VLAN_ACT_PUSH && !id_set) {
 		fprintf(stderr, "id needs to be set for %s\n",
 			action_names[action]);
 		explain();
 		return -1;
 	}
 
+	if (action == TCA_VLAN_ACT_MODIFY &&
+	   (!id_set && !prio_set && !proto_set)) {
+		fprintf(stderr, "id, prio or proto needs to be set for %s\n",
+			action_names[action]);
+		explain();
+		return -1;
+	}
+
 	if (action == TCA_VLAN_ACT_PUSH_ETH) {
 		if (!dst_mac_set) {
 			fprintf(stderr, "dst_mac needs to be set for %s\n",
